plutus-pab/src/Cardano/Protocol/Socket/Client.hs

```diff
@@ -28,62 +28,43 @@ import           Ouroboros.Network.Snocket
 import           Ouroboros.Network.Socket
 
 import           Cardano.Protocol.Socket.Type
-import           Ledger                                              (Block, Tx (..))
+import           Ledger                                              (Block, Slot (..), Tx (..))
 
 -- | Client handler, used to communicate with the client thread.
 data ClientHandler = ClientHandler {
-        chOutputQueue :: TQueue Block,
-        chInputQueue  :: TQueue Tx,
-        chSocketPath  :: FilePath,
-        chThreadId    :: ThreadId
+        chInputQueue :: TQueue Tx,
+        chSocketPath :: FilePath,
+        chThreadId   :: ThreadId
     }
 
--- | Flush all the blocks that have accumulated on the queue.
---   It does not block.
-flushBlocks ::
-    ClientHandler
- -> IO [Block]
-flushBlocks (ClientHandler outputQueue _ _ _) = go []
-    where
-      go :: [Block] -> IO [Block]
-      go acc = atomically (tryReadTQueue outputQueue) >>= \case
-          Nothing    -> pure (reverse acc)
-          Just block -> go   (block : acc)
-
--- | Read the next block. It will block if no new block is
---   available.
-getBlock ::
-    ClientHandler
- -> IO Block
-getBlock (ClientHandler outputQueue _ _ _) =
-    atomically (readTQueue outputQueue)
-
 -- | Queue a transaction to be sent to the server.
 queueTx ::
     ClientHandler
  -> Tx
  -> IO ()
-queueTx (ClientHandler _ inputQueue _ _) tx =
+queueTx (ClientHandler inputQueue _ _) tx =
     atomically (writeTQueue inputQueue tx)
 
 -- | Forks and starts a new client node, returning the newly allocated thread id.
 runClientNode :: FilePath
+              -> (Block -> Slot -> IO ())
```

Comment:
Why the callback instead of the queue?

Reply:
It seems more powerful than using a queue and it is more faithful to the idea that we are trying to use as few queues as possible. It is also quite possible to use the callback to implement something based on a queue, whereas switching from the queue to a callback requires starting a new thread.

Reply:
ok, makes sense. What happens if the callback hasn't finished when the next block comes in?

Reply:
Synchronisation happens inside whatever callback you are using. In this case it's synchronised on the `MVar AppState` from `ChainIndex`.